### PR TITLE
[WIP] Rspace/rchain 3068/write-through cache

### PR DESCRIPTION
**Pull request by dzajkowski**
## Overview
<sup>_What this PR does, and why it's needed_</sup>

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by dzajkowski** [comment]
bors try

**Comment by dzajkowski** [comment]
bors try

**Comment by dzajkowski** [comment]
bors try

File: rspace/src/main/scala/coop/rchain/rspace/LMDBStore.scala [hunk 1]
```diff
@@ -139,24 +159,30 @@ class LMDBStore[F[_], C, P, A, K] private[rspace] (
       txn: Transaction,
       channels: Seq[C],
       continuation: WaitingContinuation[P, K]
-  ): Unit =
+  ): Unit = {
+    val cloned       = LMDBStore.roundTrip(continuation)
```
Comment (lukasz-golebiewski) [inline rspace/src/main/scala/coop/rchain/rspace/LMDBStore.scala]: Why is this needed?

File: rspace/src/main/scala/coop/rchain/rspace/LMDBStore.scala [hunk 2]
```diff
@@ -216,8 +244,10 @@ class LMDBStore[F[_], C, P, A, K] private[rspace] (
           val newJoins = removeFirst(joins)(_ == channels)
           if (newJoins.nonEmpty)
             insertJoin(txn, joinedChannelHash, removeFirst(joins)(_ == channels))
-          else
+          else {
             _dbJoins.delete(txn, joinedChannelHash)
+            joinCache.remove(joinedChannelHash)
```
Comment (lukasz-golebiewski) [inline rspace/src/main/scala/coop/rchain/rspace/LMDBStore.scala]: I think you should use `.remove(joinedChannelHash, joins)` which is atomic

File: rspace/src/main/scala/coop/rchain/rspace/LMDBStore.scala [hunk 3]
```diff
@@ -216,8 +244,10 @@ class LMDBStore[F[_], C, P, A, K] private[rspace] (
           val newJoins = removeFirst(joins)(_ == channels)
           if (newJoins.nonEmpty)
             insertJoin(txn, joinedChannelHash, removeFirst(joins)(_ == channels))
-          else
+          else {
             _dbJoins.delete(txn, joinedChannelHash)
```
Comment (lukasz-golebiewski) [inline rspace/src/main/scala/coop/rchain/rspace/LMDBStore.scala]: If delete throws is the cache going to contain a stale entry?

File: rspace/src/main/scala/coop/rchain/rspace/LMDBStore.scala [hunk 4]
```diff
@@ -295,6 +327,12 @@ class LMDBStore[F[_], C, P, A, K] private[rspace] (
 
 object LMDBStore {
 
+  def roundTrip[K: Codec](k: K): K =
```
Comment (lukasz-golebiewski) [inline rspace/src/main/scala/coop/rchain/rspace/LMDBStore.scala]: There's a similar implementation in `InMemoryStore`. Let's move to a common location and re-use

File: rspace/src/main/scala/coop/rchain/rspace/LMDBStore.scala [hunk 5]
```diff
@@ -55,38 +61,52 @@ class LMDBStore[F[_], C, P, A, K] private[rspace] (
       txn: Transaction,
       channelsHash: Blake2b256Hash
   ): Option[GNAT[C, P, A, K]] =
-    _dbGNATs.get(txn, channelsHash)(codecGNAT[C, P, A, K])
+    gnatCache.get(channelsHash).orElse {
+      _dbGNATs.get(txn, channelsHash)(codecGNAT[C, P, A, K]).map { v =>
+        gnatCache.put(channelsHash, v)
```
Comment (lukasz-golebiewski) [inline rspace/src/main/scala/coop/rchain/rspace/LMDBStore.scala]: `putIfAbsent` is atomic

File: rspace/src/main/scala/coop/rchain/rspace/LMDBStore.scala (hunk 1 above)
Comment (dzajkowski) [reply rspace/src/main/scala/coop/rchain/rspace/LMDBStore.scala]: because there is no guarantee that the continuation wont have internal state
